tests/src/comparators/test_byGivenOrder.cpp

```diff
@@ -0,0 +1,54 @@
+/* test_byGivenOrder.cpp
+Copyright (c) 2022 by Michael Zahniser
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "es-test.hpp"
+
+// Include only the tested class's header.
+#include "../../source/comparators/ByGivenOrder.h"
+
+// ... and any system includes needed for the test file.
+
+namespace { // test namespace
+
+// #region unit tests
+SCENARIO( "Test basic ByGivenOrder functionality." , "[ByGivenOrder]" ) {
+	GIVEN("An order") {
+		const std::vector<int> givenOrder = { 4, 2, 8, 6 };
+		ByGivenOrder<int> comparator(givenOrder);
+
+		THEN( "Known elements are sorted by the given order" ) {
+			std::vector<int> toSort = { 2, 4, 6 };
+			const std::vector<int> expectedOrder = { 4, 2, 6 };
+			std::sort(toSort.begin(), toSort.end(), comparator);
+			REQUIRE( toSort == expectedOrder );
+		}
+
+		THEN( "Unknown elements are sorted by their native order" ) {
+			std::vector<int> toSort = { 5, 1, 3 };
+			const std::vector<int> expectedOrder = { 1, 3, 5 };
+			std::sort(toSort.begin(), toSort.end(), comparator);
+			REQUIRE( toSort == expectedOrder );
+		}
+
+		THEN( "Unknown elements are sorted after known elements" ) {
+			std::vector<int> toSort = { 2, 4, 6, 8, 5, 1, 3 };
+			const std::vector<int> expectedOrder = { 4, 2, 8, 6, 1, 3, 5 };
+			std::sort(toSort.begin(), toSort.end(), comparator);
+			REQUIRE( toSort == expectedOrder );
+		}
```

Comment:
The tests here look good to me.
Adding some tests with a givenOrder that has duplicates might be good for test-coverage (what if the number 4 appears on location 1 and on location 3?). And adding some tests where the toSort has duplicate numbers might also be good for test-coverage. But I consider none of those blocking for merging.

Reply:
> Adding some tests with a givenOrder that has duplicates might be good for test-coverage (what if the number 4 appears on location 1 and on location 3?).

That's not a valid way to use ByGivenOrder. I guess it could raise an exception, or log an error, or something.

> And adding some tests where the toSort has duplicate numbers might also be good for test-coverage.

Added explicit tests of comparing something to itself.

Reply:
I'm fine with the current behavior (any occurrence beyond the first is silently ignored).

It's possible we use this with user-controlled data where duplicates can occur, so throwing a runtime error would be undesirable.